### ISSUE TITLE
Commanline MSTest does not see xUnit's tests

### ISSUE THREAD

**Issue by quetzalcoatl** [quetzalcoatl/xvsr10#2]
Once upon a time, I've got an email from EmperorXLII about a bug/missing feature. I've investigated it back then, and finally I think it's worth to record the findings finally here:

> I don't have time to create a github account now to report the issue,
> but I did want to let you know that the new .vsix does not support
> command-line execution through MSTest.exe of test assemblies.
> 
> In other words, from a VS command prompt, "mstest
> /testcontainer:MyTest.dll" reports "No tests to execute".
> 
> I still had the .reg file from the previous version; adding that to
> the registry fixes the mstest issue.

**Comment by quetzalcoatl** [comment]
My original response:

> It seems that MSTest differs heavily from VS internal runner. The VS scans per-user and per-instance HKCU mostly, while MSTest relies on global HKLM. VSIX is installed only into HKCU, so the MSTest will never see anything that extensions has provided. Of course, the MSTest doesn't seem to scan the VS's extension directory, so even if the reg keys were patched, the process does would not see the extension's binaries and test enumeration would fail on very early stages simply due to "test TIP not found" error.
> 
> If it started working when you used old reg file, it means that you have probably activated an version older than 2.1 (older than the one with 'installer'), because the mstest saw some TIP implementation. I'm almost sure you've had the 1.2 or 2.0 version in the /PrivateAssemblies, and that the .reg file activated it in parallel to those current in /PrivateAssemblies/xunit.runner.(...) directory. 
> 
> Repairing the assembly lookup is trivial, as mstest.exe.config can be patches just like qtagent's, but for the registry I don't see any way except hardcoded copying from hkcu to hklm :/
> 
> Of course I'm completely against the idea of writing to global HKLM, and I dont like writing to registry at all, as the to get the vsix nicely register and unregisters itself automatically, but I've checked thorougly and the MSTest (and probably any other small tools sitting in common/ide directory) are simply unable to read the HKCU personal entries. IMHO, this is a one huge bug in the toolset. It is possible to extend the mstest.exe in a per-user basis, but it literally requires rewriting the mstest.exe with a heavily patched custom version, and - more to it - the patches really would have to be done at the lower-than-native level, as the bug sits in the native dll import tables...
> 
> Until I have tons of time to reverse the hijacks from devenv.exe, the runner will probably stick to writing to HKLM :/

**Comment by quetzalcoatl** [comment]
Gory details: http://quetzalcoatl-pl.blogspot.com/2012/04/visualstudio-2010-registry-chaos.html
scratchpad: https://github.com/quetzalcoatl/xvsr10/tree/iss0002-mstest-cmdline

**Comment by quetzalcoatl** [comment]
Quick info: I've already tried few tricks and it turned out that doing it in a friendly and clean way is simply impossible. When writing the 'installer' I strived to make it as your-computer-friendly as possible and introduce the least changes as possible. Here, either the root VS configuration located in HKLM would have to be 'contaminated' (which I rather despise), or the mstest.exe have to be completely replaced with a transparently patched version that would provide correct 'shadowing' of the registry branches, just like the devenv.exe does it (and I dont know exactly how to implement that detours yet).

And the current status: impossible/pending (please note that this is not a 'wontfix' :) )

For anybody not minding pollutiong their HKLM, I can glue up some .reg files that will "just force" the MSTest to see the extensions. Just drop me a note, and I'll create and post them. If demand is "high enough", that patch could eventiually be built in into the installer, too.
